Added parker, weather_current, and weather_prediction drivers; Added …

akshephard opened the pull request:
…corresponding ingester plugins for each driver; Added proto files for each driver

gtfierro commented:
A couple things:

- Can you import the proto files you've created in the main `xbos.proto` file at https://github.com/gtfierro/xboswave/blob/master/proto/xbos.proto ?
- What is a "Parker"?
- Can you clarify the difference between the `weather_current` driver in the PR vs the `dark_sky` driver that's already in the repository?

anandkp92 commented:
- Parker is a refrigeration controller (https://sporlanonline.com/literature/100/cp/100-50-6.5.pdf): for controlling refrigerator/freezers

akshephard commented:
The weather_current driver is only for the current weather station observations and weather_prediction is for hourly forecast. The weather_current driver has more fields than the existing DarkSky and so does the weather_prediction. We may want to poll the DarkSky API at different rates for current and for hourly since it seems like it only updates once an hour for the predictions. I thought it made it more clear to separate them and make the name more self explanatory.

gtfierro commented on ingester/plugins/weather_prediction_plugin.go:
```diff
@@ -0,0 +1,325 @@
+            var step int
+            step = 0
+
+            //Iterate through each hour of prediction from current to 48 hours from current
```
Can you run `go fmt` on this code so its consistently formatted?

gtfierro commented on ingester/plugins/weather_prediction_plugin.go:
```diff
@@ -0,0 +1,325 @@
+
+            	if prediction.PrecipIntensity != nil {
+                    //This will be the value that is put into a field in Influx
+            		extracted.Values = append(extracted.Values, float64(prediction.PrecipIntensity.Value))
```
Have you thought about using more generic functions like we have above for this? For example [the existing weather plugin](https://github.com/gtfierro/xboswave/blob/master/ingester/plugins/weather_station_plugin.go#L82-L88). That should reduce duplicated code

gtfierro commented on proto/weather_current.proto:
```diff
@@ -0,0 +1,58 @@
+syntax = "proto3";
+package xbospb;
+import "nullabletypes.proto";
+message Weather_Current_State {
```
Can we integrate this with the existing `Weather_Station` proto file instead of creating a new one? They should be easily unified and field should be named very similarly (especially since they are based on the same data source)

gtfierro commented on python/pyxbos/pyxbos/drivers/weather_prediction/parker.proto:
```diff
@@ -0,0 +1,189 @@
+syntax = "proto3";
+package xbospb;
+import "nullabletypes.proto";
+message ParkerState {
```
Can you remove the `.proto` file from the Python package? I like to keep all of them in a single standard place (`/proto`)

gtfierro commented on python/pyxbos/pyxbos/drivers/weather_prediction/weather_prediction.proto:
```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
```
Similarly, we should remove this `.proto` file

akshephard commented:
I added a function and ran the formatting tool on the plugin. Let me know what you think of the new changes to the plugin.

gtfierro commented:
The formatting + plugin look ok to me. Have you run tests on them to make sure they behave as you expect?

I think we're almost good to merge this. The only item left is merging the different weather messages together so that we don't have two "current weather" and two "weather prediction" messages.

akshephard commented:
Alright I have removed the existing weather message from iot.proto and added both of the current and prediction message into weather_station.proto. I included that file in iot.proto and tested all of this on my fork with the NUC with everything working.

gtfierro commented:
Thanks! I think we should change the message names to `WeatherStation` and `WeatherStationPrediction` so that the prediction message doesn't require a list of `Weather_Current_State`. 

This will require altering the messages in your driver and the existing DarkSky driver and regenerating the Python files from the proto

akshephard commented:
Okay I updated our fork to your latest code and ran the drivers on the NUC. Everything works including the dark_sky driver with updated message names in weather_station.proto. The import statement in the driver had to be updated with your latest changes.

gtfierro commented:
This looks great. Thanks for your prompt changes :)

I made a quick merge fix on the DarkSky drivers, but nothing complex. I think this is good to merge!

gtfierro commented:
Just a heads up I finally got to play around with this patch a bit more and there's actually still a few issues I need to work out with it. First of all is to never *remove* entries from the proto files! This breaks all existing code. I'm going to need to move things around, so be sure that you pull/update/rebuild everything later today including the ingester plugins so that you are up to date

gtfierro commented:
Accumulating the changes in https://github.com/gtfierro/xboswave/tree/protofix so you are up to date. We really need to set up some integration/build testing to help us avoid these kinds of issues going forward
